File: DSAlgorithms/Leetcode/ts/src/arrays_hash/347-top_k_frequent.ts
const topKFrequent = (nums: number[], k: number): number[] => {
    let arr_size = nums.length;
    const hash = new Map<number, number>();
    for(let i = 0; i < arr_size; i++)
    {
        if(hash.has(nums[i]))
        {
            let current_val = hash.get(nums[i])
            hash.set(nums[i], current_val + 1);
        }
        else 
            hash.set(nums[i], 1);
    }
    const buckets: number[][] = new Array(arr_size + 1);
    for(let i = 0; i < arr_size + 1; i++)
        buckets[i] = new Array();
    hash.forEach((value, key) => {
            buckets[value].push(key);
    });
    const result: number[] = [];
    for(let i = arr_size; i >= 0; i--)
    {
        if(result.length >= k)
            break;
        if(buckets[i].length > 0)
            result.push(...buckets[i]);
    }
    return result;

};

File: DSAlgorithms/Leetcode/arrays/arrays_hash/347-top_k_frequent.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int arr_size = nums.size();
        std::unordered_map<int, int> hash;
        for(int i = 0; i < arr_size; i++)
            hash[nums[i]]++;
        
        std::vector<std::vector<int>> buckets(arr_size + 1);
        for(auto it = hash.begin(); it!= hash.end(); it++)
            buckets[it->second].push_back(it->first);
        std::vector<int> result;
        for(int i = arr_size; i >= 0; i--)
        {
            if(result.size() >= k)
                break;
            if(!buckets[i].empty())
                result.insert(result.end(), buckets[i].begin(), buckets[i].end());
        }
        return result;
    }
};